Translate OGGBundle specification to english. (#7809)
4teamwork/opengever.core: docs/public/dev-manual/oggbundle/index.rst

| 0.1.2         | 16.01.2017   | LG          | JSON schemas referenced                                |
+---------------+--------------+-------------+--------------------------------------------------------+
| 0.1.1         | 12.01.2017   | LG          | Not allowed file formats defined                       |
+---------------+--------------+-------------+--------------------------------------------------------+
| 0.1           | 26.11.2016   | LG, DE      | Initial draft                                          |
+---------------+--------------+-------------+--------------------------------------------------------+

Status: in progress

The interface described here is used for the one-time import of a repository, its filing positions, dossiers/subdossiers and documents/mails into OneGov GEVER. The migration takes place from a JSON-based intermediate format. This must correspond to a valid schema and the data it contains must comply with the business rules applicable in OneGov GEVER.

Importable content types
------------------------

+------------------------------+-------------+

> | **Repositories**             | Ja          |

Reviewer: ```suggestion
| **Repositories**             | Yes         |
```

+------------------------------+-------------+

> | **Repository folders**       | Ja          |

Reviewer: ```suggestion
| **Repository folders**       | Yes         |
```

+------------------------------+-------------+

> | **Workspace roots**          | Ja          |

Reviewer: ```suggestion
| **Workspace roots**          | Yes         |
```

+------------------------------+-------------+

> | **Workspaces**               | Ja          |

Reviewer: ```suggestion
| **Workspaces**               | Yes         |
```

+------------------------------+-------------+

> | **Workspace folders**        | Ja          |

Reviewer: ```suggestion
| **Workspace folders**        | Yes         |
```

+------------------------------+-------------+

> | **Dossiers**                 | Ja          |

Reviewer: ```suggestion
| **Dossiers**                 | Yes         |
```

+------------------------------+-------------+

> | **Documents**                | Ja          |

Reviewer: ```suggestion
| **Documents**                | Yes         |
```

+------------------------------+-------------+

> | **Mails**                    | Ja          |

Reviewer: ```suggestion
| **Mails**                    | Yes         |
```

+------------------------------+-------------+

> | **OGDS Users**               | Ja          |

Reviewer: ```suggestion
| **OGDS Users**               | Yes         |
```

+------------------------------+-------------+

> | Contacts                     | Nein \*     |

Reviewer: ```suggestion
| Contacts                     | No \*       |
```

+------------------------------+-------------+

> | Org units                    | Nein        |

Reviewer: ```suggestion
| Org units                    | No          |
```

+------------------------------+-------------+

> | Meetings                     | Nein        |

Reviewer: ```suggestion
| Meetings                     | No          |
```

+------------------------------+-------------+

> | Tasks / forwardings          | Nein        |

Reviewer: ```suggestion
| Tasks / forwardings          | No          |
```

All path specifications in a bundle are relative to the root of the bundle.

The bundle consists of a collection of JSON files whose contents must follow a specific schema, and a subdirectory :ref:`files/ <files>` which contains the files for documents (primary data).

----

A bundle contains one file per content type to be imported. In it, the respective content must be stored flat (without nesting) in JSON format. For each such file, a `JSON schema <http://json-schema.org/>`__ is provided, which precisely describes the structure of the JSON file and with which the content must be validated before an import. The following sections describe the content types currently supported and the associated files in the bundle.

|img-image-1|

Configuration and bundle metadata

> ----------------------------------

Reviewer: ```suggestion
---------------------------------
```

metadata.json
~~~~~~~~~~~~~

This file contains metadata about the bundle, e.g. the creation date and creator of the bundle or the intended use (optional).

configuration.json
~~~~~~~~~~~~~~~~~~

This file contains the configuration of the client, especially the value ranges needed to validate the contents, which are configurable for certain fields per client.

JSON schema: :ref:`configuration.schema.json <configuration_schema_json>`

Data for content types
----------------------

reporoots.json
~~~~~~~~~~~~~~

> This file contains one or more repository roots.

Reviewer: evtl. classification system roots

JSON Schema: :ref:`reporoots.schema.json <reporoots_schema_json>`

repofolders.json
~~~~~~~~~~~~~~~~

> This file contains the individual order items that are stored in the order system roots.

Reviewer: ```suggestion
This file contains the individual classification items that are stored in the classification system root.
```
Ich weiss nicht welche Begriffe wir im Englischen für Ordnungssystem verwenden. Allenfalls auch repository folder und repository root.

JSON Schema: :ref:`repofolders.schema.json <repofolders_schema_json>`

workspaceroots.json
~~~~~~~~~~~~~~~~~~~

This file contains a workspace root.

> If a workspace root already exists on the installation into which an OGGBundle with workspaces is imported, this file can be omitted. During the import, it is then assumed that exactly one workspace root already exists and the workspace are imported into this workspace root.

Reviewer: ```suggestion
If a workspace root already exists on the installation into which an OGGBundle with workspaces is imported, this file can be omitted. During the import, it is then assumed that exactly one workspace root already exists and the workspaces are imported into this workspace root.
```

In this case, the workspace must not have a ``parent_guid`` set.

JSON Schema: :ref:`workspaceroots.schema.json <workspaceroots_schema_json>`

workspaces.json
~~~~~~~~~~~~~~~

This file contains one or more workspaces.

The workspaces are assigned to a workspace root via the ``parent_guid``, which is also included in the bundle.

Alternatively, the ``parent_guid`` for workspaces, and the definition of a workspace root in the ``workspaceroots.json`` can be omitted - the workspaces will then be imported into an already existing workspace root.

JSON schema: :ref:`workspaces.schema.json <workspaces_schema_json>`

workspacefolders.json
~~~~~~~~~~~~~~~~~~~~~

This file contains one or more workspaces folders.

JSON Schema: :ref:`workspacefolders.schema.json <workspacefolders_schema_json>`

dossiers.json
~~~~~~~~~~~~~

> This file contains dossiers and subdossiers, these can be stored in the folder positions.

Reviewer: ```suggestion
This file contains dossiers and subdossiers, these can be stored in the repository folder (classification items).
```

The actual title / file name used in OneGov GEVER is controlled by the **title** attribute in the metadata supplied in **documents.json**: In the **title** attribute, the original file name, including file extension, should be supplied. In OneGov GEVER, the title of the document is then derived from this attribute by removing the file extension. The file extension itself, on the other hand, is used to determine the content type (MIME type).

The following file types are not allowed in OGGBundles:

- **.exe**

- **.dll**

Paths / file names may only contain alphanumeric characters, underscore and hyphen (**[0-9][a-zA-Z][-\_]**). All paths are case-sensitive, and must not exceed a maximum length of 255 characters. The paths are to be specified as UNIX paths relative to the root of the bundle (separated with forward slash).

Mapping of nesting (containment)

> -------------------------------------------

Reviewer: ```suggestion
--------------------------------
```

parent_reference
~~~~~~~~~~~~~~~~

As an alternative to the GUID, the reference number of an object can also be used as a unique reference to the parent. The use of the reference number as a parent pointer allows already existing objects to be referenced via their unique reference number, thus enabling partial imports. For example, it is possible to import documents into an existing dossier by referencing this dossier via its reference number.

If the reference number is used for referencing, the attribute **parent\_reference** (instead of **parent\_guid**) must be set. The file number in this attribute is expected to be a nested array of integers that map the individual components of the file number (without formatting). Example: `[[1, 3, 5], [472, 9]` corresponds to the reference number `1.3.5 / 472.9` (position 1.3.5, dossier 472, subdossier 9):

.. code::

  {

>   "guid": [ "9999-0000-0000" ],

Reviewer: ```suggestion
  "guid": "9999-0000-0000-0000",
```

  "parent_reference": [[1, 3, 5], [472, 9],
  ...
  }

See also section :ref:`Business rules <business_rules>` for details of which content types may be nested and how.

Permissions
-----------

Permissions are inherited by default to children in OneGov GEVER. It is permitted to set permissions at the level of the repositoryroot, repositoryfolder and dossier, whereby permissions at the level of the dossier should be the exception.

> The possible permissions are basically dependent on the respective content type. The concrete permitted values can be taken from the JSON schema for the type. For most GEVER content, however, the controllable permissions are identical - the exception being workspace content.

Reviewer: ```suggestion
The possible permissions are basically dependent on the respective content type. The permitted values can be taken from the JSON schema for the type. For most GEVER content, however, the controllable permissions are identical - the exception being workspace content.
```

    }
  }

Permissions can be assigned granularly for the following roles:

- ``read`` (read)

- ``add`` (add dossiers)

- ``edit`` (edit dossiers)

- ``close`` (close dossiers)

- ``reactivate`` (reactivate dossiers)

- ``manage_dossiers`` (manage dossiers)

> In addition, a **block\_inheritance** flag can be used to specify whether the inheritance of permissions should be interrupted at this level. This means that from this level onwards, only the explicitly defined access authorisations are valid and no authorisations are taken over from the parent via inheritance:

Reviewer: ```suggestion
In addition, a **block\_inheritance** flag can be used to specify whether the inheritance of permissions should be interrupted at this level. This means that from this level onwards, only the explicitly defined access authorizations are valid and no authorizations are taken over from the parent via inheritance:
```
Ich denke wir verwenden generell die Amerikanische Schreibweise.

| 

+------------------------------+------------------+
| ``dossier-state-active``     | Active           |
+------------------------------+------------------+
| ``dossier-state-resolved``   | Resolved         |
+------------------------------+------------------+

Initial state: ``dossier-state-active``

Therefore, to deliver a dossier in the completed state, the
``review_state`` is set to the appropriate value:

  ...

  "``review_state``: ``dossier-state-resolved``,

  ...

> When a dossier is delivered in the resolved state, each contained subdossier SHALL also have the status ``dossier-state-resolved``. Fulfilment of the rules on ``loose sheets`` and date ranges, on the other hand, is recommended but not strictly required for import (will be logged but imported ``as-is``).

Reviewer: ```suggestion
When a dossier is delivered in the resolved state, each contained subdossier MUST also have the status ``dossier-state-resolved``. Fulfilment of the rules on ``loose sheets`` and date ranges, on the other hand, is recommended but not strictly required for import (will be logged but imported ``as-is``).
```

JSON Schema: :ref:`dossiers.schema.json <dossiers_schema_json>`

Documents
~~~~~~~~~
| 

+----------------------------+----------------------+
| ``document-state-draft``   | (default state)      |
+----------------------------+----------------------+

Inital state: ``document-state-draft``

JSON Schema: :ref:`documents.schema.json <documents_schema_json>`

Creator
-------

The creator of an object can be set for all contents with the property ``_creator``. The corresponding journal entries are also recorded in the name of the creator of the respective object.

> Redirects to previous threads

Reviewer: ```suggestion
Redirects to previous URL paths
```

- Closed dossiers:

   - Closed dossiers may not contain open subdossiers.

   - If a dossier is closed and has subdossiers, all documents must be assigned to a subdossier, the main dossier must not contain any documents directly assigned to it ("no loose sheets").

   - The end date of a resolved dossier must always be greater than or equal to the end date of all its subdossiers, and greater than or equal to the document date of a contained document.

- A repositoryfolder can only contain either dossiers or further repositoryfolders, never objects of both content types at the same time. Accordingly, dossiers may only be contained in leaf nodes of the repository.

- For the following fields, the choice is restricted by the parent:

   - ``custody_period``

   - ``archival_value``

   - ``classification``

   - ``privacy_layer``

   - ``retention_period`` - *Depending on the configuration, this rule may not be active*.

   Restricting in this context means that the list of available
   available elements according to the JSON schema definition to the element

>    selected by the parent and all subsequent elements.

Reviewer: ```suggestion
   selected by the parent and all subsequent elements
```
